core/src/com/unciv/scripting/ScriptingBackend.kt

```diff
@@ -0,0 +1,300 @@
+        if (' ' in command) {
+            return AutocompleteResults(true, listOf(), getCommandHelpText(command.split(' ')[0]))
+        } else {
+            return AutocompleteResults(false, commandshelp.keys.filter({ c -> c.startsWith(command) }).map({ c -> c + " " }), "")
+        }
+    }
+
+    override fun exec(command: String): String {
+        var args = command.split(' ')
+        var out = "\n> ${command}\n"
+        fun appendOut(text: String) {
+            out += text + "\n"
+        }
+        when (args[0]) {
+            "help" -> {
+                if (args.size > 1) {
+                    appendOut(getCommandHelpText(args[1]))
+                } else {
+                    appendOut(commandshelp.keys.joinToString(", "))
+                }
+            }
+            "countcities" -> {
+                if (!(scriptingScope.isInGame)) { appendOut("Must be in-game for this command!"); return out }
+                appendOut(scriptingScope.civInfo!!.cities.size.toString())
+            }
+            "locatebuildings" -> {
+                var buildingcities:List<String> = listOf()
+                if (!(scriptingScope.isInGame)) { appendOut("Must be in-game for this command!"); return out }
+                if (args.size > 1) {
+                    var searchfor = args.slice(1..args.size-1).joinToString(" ").trim(' ')
+                    buildingcities = scriptingScope.civInfo!!.cities
+                        .filter {
+                            searchfor in it.cityConstructions.builtBuildings ||
+                            it.cityConstructions.builtBuildings.any({ building ->
+                                scriptingScope.gameInfo!!.ruleSet.buildings[building]!!.requiresResource(searchfor)
+                            })
+                        }
+                        .map { it.name }
+                }
+                appendOut(buildingcities.joinToString(", "))
+            }
+            "missingbuildings" -> {
+                var buildingcities:List<String> = listOf()
+                if (!(scriptingScope.isInGame)) { appendOut("Must be in-game for this command!"); return out }
+                if (args.size > 1) {
+                    var searchfor = args.slice(1..args.size-1).joinToString(" ").trim(' ')
+                    buildingcities = scriptingScope.civInfo!!.cities
+                        .filter { !(searchfor in it.cityConstructions.builtBuildings) }
+                        .map { it.name }
+                }
+                appendOut(buildingcities.joinToString(", "))
+            }
+            "listcities" -> {
+                if (!(scriptingScope.isInGame)) { appendOut("Must be in-game for this command!"); return out }
+                appendOut(scriptingScope.civInfo!!.cities
+                    .map { city -> city.name }
+                    .joinToString(", ")
+                )
+            }
+            "cheatson" -> {
```

Comment:
Would it be possible to disable this command in multiplayer games?

Reply:
Hm. Good point. Though I mostly mean for the built-in commands to be a demo until I or someone else cooks up a good cross-platform IPC system for a Python/Lua/JS interpreter, and I've also added another backend that gives near-full reflective access to the entire program state, so I'd prefer to not to set a precedent of implementing too much functionality like permissions in the command parsers themselves. IMO permissions and action legality should be handled on the internal API level, ideally in a way that can eventually be unified with the main game GUI.

I'll think about starting to implement a `CheatAPI` controller in the command backend scope, or maybe just disable the console screen entirely on multiplayer.

Reply:
@xlenstra Right, finally did this as below:

https://github.com/yairm210/Unciv/blob/f1abea3dfe41c5c1227759347ac83fe6f15af84d/core/src/com/unciv/scripting/ScriptingState.kt#L129-L133

This is at the entrypoint for every script/command that can/should get run, so it should blanket block everything with a nice notification. A further check will have to prevent the game from trying to automatically run/load mod script handlers (which will fail), once those are implemented. I'm not worried about pass-and-play, where there's easier ways both to cheat and be caught cheating anyway.

However, can you confirm that `UncivGame.Current.gameInfo.gameParameters.isOnlineMultiplayer` is the right check to make here?